### PR TITLE
Execution event cluster validation

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Allow modifying the cluster state for queued workflow execution events and validate all other events come from the expected cluster.

---
**NOTE**

We could optimize this call by adding a cluster column for node and task executions in the db, i'm not sure if that's entirely necessary though

---

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1929

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/manager/impl/validation/shared_execution.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		Name:    executionID.Name,
+	})
+	if err != nil {
+		logger.Debugf(ctx, "Failed to find existing execution with id [%+v] with err: %v", executionID, err)
+		return err
+	}
+	if workflowExecution.Cluster != cluster {
+		errorMsg := fmt.Sprintf("Cluster/producer from event [%s] does not match existing workflow execution cluster: [%s]",
+			workflowExecution.Cluster, cluster)
+		return errors.NewIncompatibleClusterError(ctx, errorMsg, workflowExecution.Cluster)
```
Comment (pmahindrakar-oss) [inline pkg/manager/impl/validation/shared_execution.go:31]: Can we also add name of the cluster it was compared against. i.e cluster string in the IncompatibleClusterError

File: pkg/manager/impl/validation/shared_execution.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+func ValidateCluster(ctx context.Context, db repositories.RepositoryInterface, executionID *core.WorkflowExecutionIdentifier, cluster string) error {
+	if len(cluster) == 0 || cluster == common.DefaultProducerID {
+		return nil
+	}
+	workflowExecution, err := db.ExecutionRepo().Get(ctx, repoInterfaces.Identifier{
+		Project: executionID.Project,
+		Domain:  executionID.Domain,
+		Name:    executionID.Name,
+	})
```
Comment (pmahindrakar-oss) [inline pkg/manager/impl/validation/shared_execution.go:23]: We do already check if execution exists here https://github.com/flyteorg/flyteadmin/blob/296655a415d36e20bdedf98e41f890eb4901480c/pkg/manager/impl/node_execution_manager.go#L92 before calling ValidateCluster.

Can ValidateCluster be made independent of the DB call inorder to reduce the same checks on the DB or probably have a common refactored Validate function which does these checks

File: pkg/manager/impl/validation/shared_execution.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+	"github.com/flyteorg/flyteadmin/pkg/common"
+	"github.com/flyteorg/flyteadmin/pkg/errors"
+	"github.com/flyteorg/flyteadmin/pkg/repositories"
+	repoInterfaces "github.com/flyteorg/flyteadmin/pkg/repositories/interfaces"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/flyteorg/flytestdlib/logger"
+)
+
+func ValidateCluster(ctx context.Context, db repositories.RepositoryInterface, executionID *core.WorkflowExecutionIdentifier, cluster string) error {
```
Comment (pmahindrakar-oss) [inline pkg/manager/impl/validation/shared_execution.go:15]: May be a doc that this validate executionId belongs to passed in cluster string.

Also can we mention somewhere why its ok to have workflowExecution.Cluster different than common.DefaultProducerID and still raise no error

**Review by pmahindrakar-oss — DISMISSED** [review]

**Review by EngHabu — DISMISSED** [review]
Looks good to me... Praful had some comments though...

File: pkg/manager/impl/validation/shared_execution.go (hunk 1 above)
Comment (katrogan) [reply pkg/manager/impl/validation/shared_execution.go:31]: that's already included in the error message defined in the line above

File: pkg/manager/impl/validation/shared_execution.go (hunk 3 above)
Comment (katrogan) [reply pkg/manager/impl/validation/shared_execution.go:15]: Already done here: https://github.com/flyteorg/flyteadmin/pull/307/files#diff-89fdbc388b31a2407358d353ba0dc95fb7b9a4862b0528fc0792aabb67e8b178R15 but I can copy the comment

File: pkg/manager/impl/validation/shared_execution.go (hunk 2 above)
Comment (katrogan) [reply pkg/manager/impl/validation/shared_execution.go:21]: sure, done

File: pkg/manager/impl/validation/shared_execution.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply pkg/manager/impl/validation/shared_execution.go]: Ok . that should be good then.

**Comment by katrogan** [comment]
PTAL @pmahindrakar-oss @EngHabu

**Review by EngHabu — APPROVED** [review]
